File: java/sdk/src/main/java/network/t0/sdk/common/ValidationUtils.java
package network.t0.sdk.common;

import build.buf.protovalidate.ValidationResult;
import build.buf.protovalidate.Violation;

import java.util.stream.Collectors;

/**
 * Shared utilities for protovalidate violation formatting.
 */
public final class ValidationUtils {

    private ValidationUtils() {}

    /**
     * Formats validation violations into a human-readable semicolon-separated string.
     */
    public static String formatViolations(ValidationResult result) {
        return result.getViolations().stream()
                .map(Violation::toProto)
                .map(v -> {
                    String fieldName = v.getField().getElementsCount() > 0
                            ? v.getField().getElements(0).getFieldName()
                            : v.getField().toString();
                    return fieldName + ": " + v.getMessage();
                })
                .collect(Collectors.joining("; "));
    }
}

File: csharp/sdk/T0.ProviderSdk/Common/ValidationUtils.cs
using ProtoValidate;

namespace T0.ProviderSdk.Common;

/// <summary>
/// Shared utilities for protovalidate violation formatting.
/// </summary>
internal static class ValidationUtils
{
    internal static string FormatViolations(ValidationResult result) =>
        string.Join("; ", result.Violations.Select(v => v.ToString()));
}
